// @simple-html/core/dist/symbols.js
export function initSymbolCache() {
    if (!globalThis._LHF_SYMBOL) {
        globalThis._LHF_SYMBOL = {};
        globalThis._LHF_PROP_SYMBOL = {};
    }
}
export function getObservedAttributesMapSymbol() {
    if (!globalThis._LHF_SYMBOL.observedAttributesMap) {
        globalThis._LHF_SYMBOL.observedAttributesMap = Symbol('observedAttributesMap');
        return globalThis._LHF_SYMBOL.observedAttributesMap;
    }
    else {
        return globalThis._LHF_SYMBOL.observedAttributesMap;
    }
}
export function getObservedAttributesSymbol() {
    if (!globalThis._LHF_SYMBOL.observedAttributes) {
        globalThis._LHF_SYMBOL.observedAttributes = Symbol('observedAttributes');
        return globalThis._LHF_SYMBOL.observedAttributes;
    }
    else {
        return globalThis._LHF_SYMBOL.observedAttributes;
    }
}
export function getPropSymbol(name) {
    if (!globalThis._LHF_PROP_SYMBOL[name]) {
        globalThis._LHF_PROP_SYMBOL[name] = Symbol(name);
        return globalThis._LHF_PROP_SYMBOL[name];
    }
    else {
        return globalThis._LHF_PROP_SYMBOL[name];
    }
}
export function getLoggerSymbol() {
    if (!globalThis._LHF_SYMBOL.logger) {
        globalThis._LHF_SYMBOL.logger = Symbol('logger');
        return globalThis._LHF_SYMBOL.logger;
    }
    else {
        return globalThis._LHF_SYMBOL.logger;
    }
}
export function getLoggerCountSymbol() {
    if (!globalThis._LHF_SYMBOL.loggerCount) {
        globalThis._LHF_SYMBOL.loggerCount = Symbol('loggerCount');
        return globalThis._LHF_SYMBOL.loggerCount;
    }
    else {
        return globalThis._LHF_SYMBOL.loggerCount;
    }
}
export function getTransmitterSymbol() {
    if (!globalThis._LHF_SYMBOL.transmitter) {
        globalThis._LHF_SYMBOL.transmitter = Symbol('transmitter');
        return globalThis._LHF_SYMBOL.transmitter;
    }
    else {
        return globalThis._LHF_SYMBOL.transmitter;
    }
}
initSymbolCache();
