// fuse-box/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var testUtils_1 = require("./compiler/testUtils");
exports.testTransform = testUtils_1.initCommonTransform;
var transformer_1 = require("./compiler/transformer");
exports.coreTransformerList = transformer_1.BASE_TRANSFORMERS;
var fusebox_1 = require("./core/fusebox");
exports.fusebox = fusebox_1.fusebox;
var plugin_angular_1 = require("./plugins/core/plugin_angular");
exports.pluginAngular = plugin_angular_1.pluginAngular;
var plugin_consolidate_1 = require("./plugins/core/plugin_consolidate");
exports.pluginConsolidate = plugin_consolidate_1.pluginConsolidate;
var plugin_css_1 = require("./plugins/core/plugin_css");
exports.pluginCSS = plugin_css_1.pluginCSS;
var plugin_css_in_jsx_1 = require("./plugins/core/plugin_css_in_jsx");
exports.pluginCSSInJSX = plugin_css_in_jsx_1.pluginCSSInJSX;
var plugin_customtransform_1 = require("./plugins/core/plugin_customtransform");
exports.pluginCustomTransform = plugin_customtransform_1.pluginCustomTransform;
var plugin_json_1 = require("./plugins/core/plugin_json");
exports.pluginJSON = plugin_json_1.pluginJSON;
var plugin_less_1 = require("./plugins/core/plugin_less");
exports.pluginLess = plugin_less_1.pluginLess;
var plugin_link_1 = require("./plugins/core/plugin_link");
exports.pluginLink = plugin_link_1.pluginLink;
var plugin_minify_html_literals_1 = require("./plugins/core/plugin_minify_html_literals");
exports.pluginMinifyHtmlLiterals = plugin_minify_html_literals_1.pluginMinifyHtmlLiterals;
var plugin_postcss_1 = require("./plugins/core/plugin_postcss");
exports.pluginPostCSS = plugin_postcss_1.pluginPostCSS;
var plugin_raw_1 = require("./plugins/core/plugin_raw");
exports.pluginRaw = plugin_raw_1.pluginRaw;
var plugin_replace_1 = require("./plugins/core/plugin_replace");
exports.pluginReplace = plugin_replace_1.pluginReplace;
var plugin_sass_1 = require("./plugins/core/plugin_sass");
exports.pluginSass = plugin_sass_1.pluginSass;
var plugin_stylus_1 = require("./plugins/core/plugin_stylus");
exports.pluginStylus = plugin_stylus_1.pluginStylus;
var plugin_web_worker_1 = require("./plugins/webworker/plugin_web_worker");
exports.pluginWebWorker = plugin_web_worker_1.pluginWebWorker;
var sparky_1 = require("./sparky/sparky");
exports.sparky = sparky_1.sparky;
